src/wallet/electrum.cpp

```diff
@@ -0,0 +1,237 @@
+#include <bitcoin/bitcoin/utility/container_sink.hpp>
+#include <bitcoin/bitcoin/utility/ostream_writer.hpp>
+#include <bitcoin/bitcoin/wallet/dictionary.hpp>
+#include "../math/external/pkcs5_pbkdf2.h"
+
+namespace libbitcoin {
+namespace wallet {
+namespace electrum {
+
+#ifdef WITH_ICU
+
+// Electrum mnemonic private constants.
+static constexpr size_t hmac_iterations = 2048;
+static const auto hmac_data = to_chunk("Seed version");
+static const std::string passphrase_prefix = "electrum";
+
+
+static uint32_t special_modulo(int index_distance, int dictionary_length)
+{
+    if (index_distance < 0)
+        return dictionary_length - (-index_distance % dictionary_length);
+
+    return index_distance % dictionary_length;
+}
+
+static data_chunk old_mnemonic_decode(const word_list& mnemonic)
+{
+    static constexpr size_t mnemonic_word_multiple = 3;
+
+    if ((mnemonic.size() < mnemonic_word_multiple) ||
+        (mnemonic.size() % mnemonic_word_multiple) != 0)
+        return {};
+
+    const auto& lexicon = language::electrum::en_v1;
+    const auto dictionary_length = lexicon.size();
+    const auto seed_size = sizeof(uint32_t) * mnemonic.size() / 3;
+
+    data_chunk seed;
+    seed.reserve(seed_size);
+    data_sink ostream(seed);
+    ostream_writer sink(ostream);
+
+    auto all_found = false;
+    for(size_t i = 0; i < mnemonic.size() / 3; i += 3)
+    {
+        const auto index_1 = find_position(lexicon, mnemonic[i]);
+        const auto index_2 = find_position(lexicon, mnemonic[i+1]);
+        const auto index_3 = find_position(lexicon, mnemonic[i+2]);
+
+        all_found = ((index_1 != -1) && (index_2 != -1) && (index_3 != -1));
+        if (!all_found)
+            break;
+
+        const uint32_t value = index_1 + dictionary_length *
+            special_modulo(index_2 - index_1, dictionary_length) +
+            dictionary_length * dictionary_length *
+            special_modulo(index_3 - index_2, dictionary_length);
+
+        sink.write_variable_little_endian(value);
+    }
+
+    if (!all_found)
+        return {};
+
+    return seed;
+}
+
+static bool is_old_seed(const word_list& mnemonic)
+{
+    // valid electrum 1.x seed lengths
+    static constexpr size_t seed_size_v1_1 = 16;
+    static constexpr size_t seed_size_v1_2 = 32;
+
+    // valid electrum 1.x mnemonic lengths
+    static constexpr size_t mnemonic_size_v1_1 = 12;
+    static constexpr size_t mnemonic_size_v1_2 = 24;
+
+    const auto seed = old_mnemonic_decode(mnemonic);
+
+    return !seed.empty() && (
+        seed.size() == seed_size_v1_1 ||
+        seed.size() == seed_size_v1_2 ||
+        mnemonic.size() == mnemonic_size_v1_1 ||
+        mnemonic.size() == mnemonic_size_v1_2);
+}
+
+static bool is_new_seed(const word_list& mnemonic, const data_slice& prefix)
+{
+    const auto seed = hmac_sha512_hash(
+        to_chunk(to_normal_nfkd_form(join(mnemonic))), hmac_data);
+
+    return (std::memcmp(seed.data(), prefix.data(), prefix.size()) == 0);
```

Comment:
Avoid memcmp as it can be unsafe depending on circumstance. It's better to use std::equal where possible (as in this case).